Test tooling: introduce test-label-analyzer cli - generate a page of quarantined tests for kubevirt/kubevirt

dhiller opened the pull request:
See kubevirt/project-infra#2679

# test-label-analyzer cli

This PR introduces a new cli for analysis of label data from tests inside a given code base.

Initial target is of course `kubevirt/kubevirt`, but might be useful for other [Ginkgo] V2 based projects as well.

# quarantined tests page

The PR also creates a periodic job that leverages the cli to generate an html page of tests in quarantine for kubevirt/kubevirt, example output here: https://storage.googleapis.com/kubevirt-prow/reports/quarantined-tests/kubevirt/kubevirt/index.html

# general usage

The cli is able to work in a more generalized style of using a test config in yaml form

```yaml
File: /tmp/analyzer-config.json                  
────────────────────────────────────────────────────────────────────
{                                                                   
    "categories": [                                                 
        {                                                           
            "name": "SSP test case evaluation",                                      
            "testNameLabelRE": ".*Console Proxy Operand Resource.*",
            "ginkgoLabelRE": ""                                     
        }                                                           
    ]                                                               
}                                                                   
```

which can be simplified to adding the flag

```
--test-name-label-re '.*Console Proxy Operand Resource.*'
```

# What does not (yet) work

Currently we can only filter by test name content, Ginkgo label filtering does not work, since Ginkgo `outline` doesn't resolve `Label` s as we use them inside `kubevirt/kubevirt`

[Ginkgo]: https://onsi.github.io/ginkgo/
[Ginkgo label]: https://onsi.github.io/ginkgo/#spec-labels

fossedihelm commented on robots/pkg/test-label-analyzer/stats.go:
```diff
@@ -0,0 +1,138 @@
+		if node.Spec {
+			stats.SpecsTotal++
+			for _, category := range config.Categories {
+				var testName string
+				for _, nodeFromPath := range parentsWithNode {
+					testName = strings.Join([]string{testName, nodeFromPath.Text}, " ")
+				}
+				if category.TestNameLabelRE.MatchString(testName) {
+					var path []*GinkgoNode
+					for _, pathNode := range parentsWithNode {
+						path = append(path, pathNode.CloneWithoutNodes())
+					}
+					stats.MatchingSpecPathes = append(stats.MatchingSpecPathes, &PathStats{
+						Path: path,
+					})
+				}
```
Thinking out loud: an improvement could be to pass a boolean to the depth call which tells if the parent already match. I mean if parent node path matches the regex => we can be sure that all the child nodes match as well.
Something like:
```
func traverseNodesRecursively(stats *TestStats, config *Config, gingkoOutline []*GinkgoNode, parents []*GinkgoNode, matched bool) {
	for _, node := range gingkoOutline {
		var parentsWithNode []*GinkgoNode
		parentsWithNode = append(parentsWithNode, parents...)
		parentsWithNode = append(parentsWithNode, node)
		if node.Spec {
			stats.SpecsTotal++
			for _, category := range config.Categories {
				if matched {
					addMatchingSpecPaths(stats, parentsWithNode)
				} else {
					var testName string
					for _, nodeFromPath := range parentsWithNode {
						testName = strings.Join([]string{testName, nodeFromPath.Text}, " ")
					}
					if category.TestNameLabelRE.MatchString(testName) {
						matched = true
						addMatchingSpecPaths(stats, parentsWithNode)
					}
				}
			}
		}
		traverseNodesRecursively(stats, config, node.Nodes, parentsWithNode, matched)
	}
}

func addMatchingSpecPaths(stats *TestStats, nodes []*GinkgoNode) {
	var path []*GinkgoNode
	for _, pathNode := range nodes {
		path = append(path, pathNode.CloneWithoutNodes())
	}
	stats.MatchingSpecPathes = append(stats.MatchingSpecPathes, &PathStats{
		Path: path,
	})
}
```
But I am not sure whether it is worth it.

dhiller replied:
I think it's not worth it since we only want to collect the Spec nodes, all child nodes of Spec nodes I have seen do have `node.Spec == false`, so I guess it's not. Let's defer this optimization until we see we need it. WDYT?

fossedihelm replied:
Sure :) As I said, I was thinking out loud

fossedihelm reviewed (commented):
Thanks @dhiller! Some comments below. In general, it seems to be a fantastic work :tada:

dhiller commented:
@fossedihelm thank you very much for your review, if you have time to TAL again, it would be much appreciated :)

fossedihelm commented on robots/pkg/test-label-analyzer/stats.go:
```diff
@@ -0,0 +1,138 @@
+	// MatchingSpecPathes is the slice of PathStats to matching specs for the collection of outlines traversed.
+	// Each PathStats inside this slice is the path to each of the matching specs defined by the Config
+	// being used. Please note that the GinkgoNode.Nodes are being removed during traversal.
+	MatchingSpecPathes []*PathStats `json:"matching_spec_pathes"`
```
nit: 
```suggestion
	MatchingSpecPaths []*PathStats `json:"matching_spec_paths"`
```
In case of the acceptance, keep in mind that we need to edit the README as well

fossedihelm reviewed (commented):
@dhiller Looks good. Just some other small comments

dhiller commented:
@fossedihelm I got it to work with bazel finally again :) Would appreciate another look from you! Thank you!

dhiller commented:
> @fossedihelm I got it to work with bazel finally again :) Would appreciate another look from you! Thank you!
> 
> EDIT: BTW I am really sorry for the size of the PR, but most of the changes are scattered upgrades to Ginkgo V2.

When looking at this, I have decided to move the updates to Gingko V2 to another PR which this one then should base upon. So I am reworking this, sorry for the noise. Will ping you when it's ready again.

dhiller commented:
I have rebased the commits and split out the Ginkgo V2 update into https://github.com/kubevirt/project-infra/pull/2726

Will rebase this one after the other one got in. I hope that makes it easier to review this one.

dhiller commented:
/hold

dhiller commented:
/unhold

I have rebased the PR onto the changes from the Ginkgo V2 update, should be better to read now.

@fossedihelm PTAL, thanks again!

fossedihelm reviewed (approved):
/lgtm
@dhiller Sorry for taking a long time.
I think PR description can be edited 
```
Note: Based on Ginkgo V2 updates from https://github.com/kubevirt/project-infra/pull/2726, requires a rebase after merge.
```
This is not required anymore :)
